File: src/gan/simplenet.cpp
//
// TorchKissAnn - A collection of tools to train various types of Machine learning
// algorithms on various types of EIS data
// Copyright (C) 2025 Carl Klemm <carl@uvos.xyz>
//
// This file is part of TorchKissAnn.
//
// TorchKissAnn is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// TorchKissAnn is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with TorchKissAnn.  If not, see <http://www.gnu.org/licenses/>.
//

#include "simplenet.h"
#include <cstddef>
#include <utility>
#include <fstream>
#include <json/value.h>

static std::pair<size_t, size_t> layerWidthFunction(size_t layer, size_t leyers, size_t input, size_t output)
{
	size_t layerIn = (input*(leyers-layer)/leyers)+output*(layer)/leyers;
	size_t layerOut = (input*(leyers-layer-1)/leyers)+output*(layer+1)/leyers;
	return std::pair<size_t, size_t>(layerIn, layerOut);
}

GANGeneratorImpl::GANGeneratorImpl(const Json::Value& node):
Net(node)
{
	init();
}

GANGeneratorImpl::GANGeneratorImpl(size_t inputSize, size_t outputSize):
Net(inputSize, outputSize, false)
{
	init();
}

void GANGeneratorImpl::init()
{
	setPurpose("gan generator for eis spectra");
	setInputLabel("RAND");
	register_module("GANGeneratorModel", model);
	std::pair<size_t, size_t> layerSizes = layerWidthFunction(0, 4, inputSize, outputSize);
	model->push_back(torch::nn::Linear(layerSizes.first, layerSizes.second));
	model->push_back(torch::nn::LeakyReLU(torch::nn::LeakyReLUOptions().negative_slope(0.2).inplace(true)));
	layerSizes = layerWidthFunction(1, 4, inputSize, outputSize);
	model->push_back(torch::nn::Linear(layerSizes.first, layerSizes.second));
	model->push_back(torch::nn::LeakyReLU(torch::nn::LeakyReLUOptions().negative_slope(0.2).inplace(true)));
	layerSizes = layerWidthFunction(2, 4, inputSize, outputSize);
	model->push_back(torch::nn::Linear(layerSizes.first, layerSizes.second));
	model->push_back(torch::nn::LeakyReLU(torch::nn::LeakyReLUOptions().negative_slope(0.2).inplace(true)));
	layerSizes = layerWidthFunction(3, 4, inputSize, outputSize);
	model->push_back(torch::nn::Linear(layerSizes.first, layerSizes.second));
	model->push_back(torch::nn::LeakyReLU(torch::nn::LeakyReLUOptions().negative_slope(0.2).inplace(true)));
	model->push_back(torch::nn::Tanh());
}

std::shared_ptr<torch::nn::Module> GANGeneratorImpl::operator[](size_t index)
{
	return model->ptr(index);
}

bool GANGeneratorImpl::saveToCheckpointDir(const std::filesystem::path& path)
{
	return true;
}

torch::Tensor GANGeneratorImpl::forward(torch::Tensor x)
{
	return model->forward(x);
}

GANDiscriminatorImpl::GANDiscriminatorImpl(size_t inputSize):
Net(inputSize, 1, false)
{
	init();
}

GANDiscriminatorImpl::GANDiscriminatorImpl(const Json::Value& node):
Net(node)
{

}

void GANDiscriminatorImpl::init()
{
	register_module("GANDiscriminatorFeature", feature);
	register_module("GANDiscriminatorClassifier", classifier);
	std::pair<size_t, size_t> layerSizes = layerWidthFunction(0, 4, inputSize, 1);
	feature->push_back(torch::nn::Linear(layerSizes.first, layerSizes.second));
	feature->push_back(torch::nn::LeakyReLU(torch::nn::LeakyReLUOptions().negative_slope(0.2).inplace(true)));
	layerSizes = layerWidthFunction(1, 4, inputSize, 1);
	feature->push_back(torch::nn::Linear(layerSizes.first, layerSizes.second));
	feature->push_back(torch::nn::LeakyReLU(torch::nn::LeakyReLUOptions().negative_slope(0.2).inplace(true)));
	layerSizes = layerWidthFunction(2, 4, inputSize, 1);
	feature->push_back(torch::nn::Linear(layerSizes.first, layerSizes.second));
	feature->push_back(torch::nn::LeakyReLU(torch::nn::LeakyReLUOptions().negative_slope(0.2).inplace(true)));
	layerSizes = layerWidthFunction(3, 4, inputSize, 1);
	classifier->push_back(torch::nn::Linear(layerSizes.first, layerSizes.second));
}

std::pair<torch::Tensor, torch::Tensor> GANDiscriminatorImpl::forwardSplit(torch::Tensor x)
{
	torch::Tensor features = feature->forward(x);
	torch::Tensor out = classifier->forward(features);
	return std::pair<torch::Tensor, torch::Tensor>(out, features);
}

torch::Tensor GANDiscriminatorImpl::forward(torch::Tensor x)
{
	return forwardSplit(x).first;
}

std::shared_ptr<torch::nn::Module> GANDiscriminatorImpl::operator[](size_t index)
{
	if(index < feature->size())
		return feature->ptr(index);
	else
		return classifier->ptr(index);
}

bool GANDiscriminatorImpl::saveToCheckpointDir(const std::filesystem::path& path)
{
	if(!std::filesystem::is_directory(path))
		std::filesystem::create_directories(path);
	if(!std::filesystem::is_directory(path))
		return false;

	Json::Value networkMetadata;
	getConfiguration(networkMetadata);
	std::ofstream networkMetadataFile;
	networkMetadataFile.open(path/"meta.json", std::ios_base::out);
	if(!networkMetadataFile.is_open())
		return false;
	Json::StreamWriterBuilder builder;
	const std::unique_ptr<Json::StreamWriter> writer(builder.newStreamWriter());
	writer->write(networkMetadata, &networkMetadataFile);

	networkMetadataFile.close();
	torch::save(shared_from_this(), path/"weights.pt");
	return true;
}

File: scripts/nets/simplenet.py

# TorchKissAnn - A collection of tools to train various types of Machine learning
# algorithms on various types of EIS data
# Copyright (C) 2025 Carl Klemm <carl@uvos.xyz>
#
# This file is part of TorchKissAnn.
#
# TorchKissAnn is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# TorchKissAnn is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with TorchKissAnn.  If not, see <http://www.gnu.org/licenses/>.

import torch


def layer_width_function(layer: int, layers: int, input: int, output: int):
	layer_in = int((input * (layers - layer) / layers) + output * (layer) / layers)
	layer_out = int((input * (layers - layer - 1) / layers) + output * (layer + 1) / layers)
	return layer_in, layer_out


class LinearBlock(torch.nn.Module):
	def __init__(self, input_size: int, output_size: int):
		super(LinearBlock, self).__init__()
		self.linear = torch.nn.Linear(input_size, output_size)
		self.activation = torch.nn.LeakyReLU(negative_slope=0.1, inplace=True)
		self.batchnorm = torch.nn.BatchNorm1d(output_size, eps=1e-3, momentum=0.1, affine=True)

	def forward(self, x):
		x = self.linear.forward(x)
		x = self.activation.forward(x)
		x = self.batchnorm.forward(x)
		return x


class SimpleNet(torch.nn.Module):
	def __init__(self, input_size: int, output_size: int, downsample_steps: int, extra_steps: int):
		super(SimpleNet, self).__init__()

		self.input_size = input_size
		self.layers = torch.nn.Sequential()

		layer_in, layer_out = layer_width_function(0, downsample_steps, input_size, input_size)
		self.layers.append(LinearBlock(layer_in, layer_out))

		for i in range(0, downsample_steps):
			layer_in, layer_out = layer_width_function(i, downsample_steps, input_size, output_size)
			self.layers.append(LinearBlock(layer_in, layer_out))
			if i == 0:
				for j in range(0, extra_steps):
					self.layers.append(LinearBlock(layer_out, layer_out))

	def forward(self, x: torch.Tensor):
		return self.layers.forward(x)
